django/contrib/staticfiles/storage.py

```diff
@@ -274,7 +282,56 @@ def path_level(name):
 
                 # and then set the cache accordingly
                 hashed_files[self.hash_key(name)] = hashed_name
+
+                yield name, hashed_name, processed, substitutions
+
+    def post_process(self, paths, dry_run=False, **options):
+        """
+        Post process the given OrderedDict of files (called from collectstatic).
+
+        Processing is actually two separate operations:
+
+        1. renaming files to include a hash of their content for cache-busting,
+           and copying those files to the target storage.
+        2. adjusting files which contain references to other files so they
+           refer to the cache-busting filenames.
+
+        If either of these are performed on a file, then that file is considered
+        post-processed.
+        """
+        # don't even dare to process the files if we're in dry run mode
+        if dry_run:
+            return
+
+        # where to store the new paths
+        hashed_files = OrderedDict()
+
+        # build a list of adjustable files
+        adjustable_paths = [
+            path for path in paths
+            if matches_patterns(path, self._patterns.keys())
+        ]
+
+        # Do a single pass first, this will post-process all files once, then we can repeat for adjustable files
+        for name, hashed_name, processed, _ in self._post_process(paths, adjustable_paths, hashed_files):
+            yield name, hashed_name, processed
```

Comment:
Currently this PR will end up yielding the same name several times over if it is an adjustable path. It will yield it for this initial pass and then for each additional pass as well.

I'm thinking it should actually only yield on the first path if it isn't an adjustable path, and then yield all adjustable paths after the final path, with their final hashed name. That'll mean there's a pause after the first pass and then all adjustable paths would be yielded at once. It's not ideal, but they can't really be yielded until all passes are done as they may still change. TBD - If a file survives a pass with the same hash, is it fully converged, or can it change in a later pass?